common/functions/src/scalars/strings/soundex.rs

```diff
@@ -0,0 +1,152 @@
+            if last.is_none() {
+                if !Self::is_uni_alphabetic(ch) {
+                    continue;
+                }
+
+                last = score;
+                result.push(ch.to_ascii_uppercase());
+            } else {
+                if !ch.is_ascii_alphabetic()
+                    || Self::is_drop(ch)
+                    || score.is_none()
+                    || score == last
+                {
+                    continue;
+                }
+
+                last = score;
+                result.push(score.unwrap());
+            }
+
+            count += 1;
+        }
+
+        // add '0'
+        if !result.is_empty() && count < 4 {
+            result.extend(vec!['0'; 4 - count])
+        }
+
+        result.into_bytes()
+    }
+
+    #[inline(always)]
+    fn number_map(i: char) -> Option<char> {
+        match i.to_ascii_lowercase() {
+            'b' | 'f' | 'p' | 'v' => Some('1'),
+            'c' | 'g' | 'j' | 'k' | 'q' | 's' | 'x' | 'z' => Some('2'),
+            'd' | 't' => Some('3'),
+            'l' => Some('4'),
+            'm' | 'n' => Some('5'),
+            'r' => Some('6'),
+            _ => Some('0'),
+        }
+    }
+
+    #[inline(always)]
+    fn is_drop(c: char) -> bool {
+        matches!(
+            c.to_ascii_lowercase(),
+            'a' | 'e' | 'i' | 'o' | 'u' | 'y' | 'h' | 'w'
+        )
+    }
+
+    // https://github.com/mysql/mysql-server/blob/3290a66c89eb1625a7058e0ef732432b6952b435/sql/item_strfunc.cc#L1919
+    #[inline(always)]
+    fn is_uni_alphabetic(c: char) -> bool {
+        ('a'..='z').contains(&c) || ('A'..='Z').contains(&c) || c as i32 >= 0xC0
+    }
+}
+
+impl fmt::Display for SoundexFunction {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "SOUNDEX")
+    }
+}
+
+impl Function for SoundexFunction {
+    fn name(&self) -> &str {
+        &*self._display_name
+    }
+
+    fn num_arguments(&self) -> usize {
+        1
+    }
+
+    fn return_type(&self, args: &[DataType]) -> Result<DataType> {
+        match args[0] {
+            DataType::Null => Ok(DataType::Null),
+            _ => Ok(DataType::String),
+        }
+    }
+
+    fn nullable(&self, _input_schema: &DataSchema) -> Result<bool> {
+        Ok(true)
+    }
+
+    fn eval(&self, columns: &DataColumnsWithField, input_rows: usize) -> Result<DataColumn> {
+        match columns[0].data_type() {
+            DataType::String => {
+                return match columns[0].column() {
+                    DataColumn::Constant(DataValue::String(Some(ref data)), _) => {
+                        let r = Self::soundex(data);
+                        Ok(DataColumn::Constant(DataValue::String(Some(r)), input_rows))
+                    }
+                    _ => Ok(DataColumn::Constant(DataValue::Null, input_rows)),
+                };
+            }
+            DataType::Null => Ok(DataColumn::Constant(DataValue::Null, input_rows)),
+            _ => {
```

Comment:
I did not know why this branch returns an empty String column.

Maybe we should cast the column into String column in previous.

```

// Null/Primitive array can cast into String column.
let column = columns[0].cast_with_type(DataType::String);
let column = column.string()?;

match column {
   Constant => {},
   Array(x) => {}
}

```